src/main/java/com/google/sps/servlets/ProjectServlet.java

```diff
@@ -0,0 +1,329 @@
+        projEntity = datastore.get(projKey);
+      } catch (Exception e) {
+        response.sendRedirect("/");
+        return;
+      }
+      ArrayList<String> owners =
+          (ArrayList<String>)projEntity.getProperty("owners");
+      ArrayList<String> editors =
+          (ArrayList<String>)projEntity.getProperty("editors");
+      String existingVis = (String)projEntity.getProperty("visibility");
+      if (!owners.contains(uEmail) && !editors.contains(uEmail) &&
+          !existingVis.equals(PUBLIC)) {
+        response.sendRedirect("/");
+        return;
+      }
+      projects.add(projEntity);
+    }
+
+    // Searching for multiple projects with various parameters
+    else {
+      String sort = request.getParameter("sort");
+      // Sorted in descending chronological order by default
+      if (isEmptyParameter(sort)) {
+        sort = DESCENDING_SORT;
+      }
+      sort = sort.toLowerCase();
+
+      Query projQuery = new Query("Project").addSort(
+          "utc", sort.equals(ASCENDING_SORT) ? Query.SortDirection.ASCENDING
+                                             : Query.SortDirection.DESCENDING);
+
+      // Add relevant filters to array based on parameters
+      ArrayList<Filter> allFilters = new ArrayList<Filter>();
+
+      String role = request.getParameter("role");
+      String visibility = request.getParameter("visibility");
+      // Global overrides visibility and role
+      Boolean global = Boolean.parseBoolean(request.getParameter("global"));
+      if (global) {
+        visibility = PUBLIC;
+        role = "viewer";
+      }
+
+      Filter ownFilter =
+          new FilterPredicate("owners", FilterOperator.EQUAL, uEmail);
+      Filter editFilter =
+          new FilterPredicate("editors", FilterOperator.EQUAL, uEmail);
+      Filter ownOrEditFilter = new CompositeFilter(
+          CompositeFilterOperator.OR, Arrays.asList(ownFilter, editFilter));
+
+      // By default, filter to only projects the User owns or edits
+      if (isEmptyParameter(role)) {
+        allFilters.add(ownOrEditFilter);
+      } else if (role.toLowerCase().equals("owner")) {
+        allFilters.add(ownFilter);
+      } else if (role.toLowerCase().equals("editor")) {
+        allFilters.add(editFilter);
+      }
+
+      // Don't filter by visibility by default
+      if (!isEmptyParameter(visibility) &&
+          (visibility.toLowerCase().equals(PUBLIC) ||
+           visibility.toLowerCase().equals(PRIVATE))) {
+        Filter visFilter =
+            new FilterPredicate("visibility", FilterOperator.EQUAL, visibility);
+        allFilters.add(visFilter);
+      }
+
+      // Don't filter by search term if not provided
+      // No partial matching/regex
+      String searchTerm = request.getParameter("search-term");
+      if (!isEmptyParameter(searchTerm)) {
+        Filter searchFilter = new FilterPredicate("name", FilterOperator.EQUAL,
+                                                  searchTerm.toLowerCase());
```

Comment:
nit: for js style guides, aligning to parenthesis is discouraged, and +4 line indentation is preferred. I'm not sure if it's the same for java style guides, so just a suggestion

Reply:
I found this in the Google Java style guide regarding line wrapping: "Indent continuation lines at least +4 spaces... Horizontal alignment is the practice of adding a variable number of additional spaces in your code with the goal of making certain tokens appear directly below certain other tokens on previous lines. This practice is permitted, but is never required by Google Style."